### PR TITLE
Fix TxQ LastLedgerSequence handling & add tests

### PR DESCRIPTION
**Pull request by ximinez**
I left the fix and the tests as separate commits so anyone interested can revert the fix commit locally and verify the one test fails.

According to [the code coverage report](https://codecov.io/github/ximinez/rippled/src/ripple/app/misc/impl/TxQ.cpp?ref=fc0b6bd8b4f74761f49f8b67eb0158fdccfdf084), the only things in TxQ left untested after this are logging, the fallbacks when the feature is disabled, the RPC command function, a few edge cases, and stuff that I can't really test without support for multiple transactions per account.

Reviewers: @nbougalis @seelabs

### CODE REVIEW

File: src/ripple/app/misc/impl/TxQ.cpp [hunk 1]
```diff
@@ -539,7 +531,7 @@ TxQ::processValidatedLedger(Application& app,
         && (!maxSize_ || keptCandidates < *maxSize_))
     {
         if (candidateIter->lastValid
-            && *candidateIter->lastValid >= ledgerSeq)
+            && *candidateIter->lastValid <= ledgerSeq)
```
Comment (seelabs) [inline src/ripple/app/misc/impl/TxQ.cpp]: Nit: `whose` is mispelled in the comment above (line 526)

File: src/ripple/app/tests/TxQ_test.cpp [hunk 2]
```diff
@@ -122,7 +122,7 @@ class TxQ_test : public TestSuite
     }
 
 public:
-    void run()
+    void testQueue()
```
Comment (seelabs) [inline src/ripple/app/tests/TxQ_test.cpp]: Not in this change set, the `return std::unique_ptr<Config const>(p.release());` on line 121 can be replaced with `return std::move(p);`. I usually point out where `return std::move` should be removed due to inhibiting RTO and just wanted to point out a case where it could be beneficially added :-)

File: src/ripple/app/misc/impl/TxQ.cpp (hunk 1 above)
Comment (ximinez) [inline src/ripple/app/misc/impl/TxQ.cpp]: Heh. Fixed.

File: src/ripple/app/tests/TxQ_test.cpp (hunk 2 above)
Comment (ximinez) [inline src/ripple/app/tests/TxQ_test.cpp]: Weird. I could swear I tried that in an earlier instance and it puked because of the `const` conversion, but it seems to work fine now.

Comment (ximinez) [inline src/ripple/app/tests/TxQ_test.cpp]: ...In fact, at least in VS, `return p;` is calling to the exact same overload. Let's see what CI says about it.

Comment (seelabs) [inline src/ripple/app/tests/TxQ_test.cpp]: Interesting. gcc 5.2 took the `return p;`. Clang 3.7 choked on it. So clang at least needs the `std::move`.

File: src/ripple/app/tests/TxQ_test.cpp [hunk 3]
```diff
@@ -360,6 +360,205 @@ class TxQ_test : public TestSuite
...
+            env.jt(noop(bob)));
+        submit(env,
+            env.jt(noop(charlie)));
+        submit(env,
+            env.jt(noop(daria)));
+        checkMetrics(env, 0, boost::none, 3, 2, 256, 500);
+
+        // Queue an item with a LastLedgerSeq.
+        submit(env,
+            env.jt(noop(alice), json("{\"LastLedgerSequence\":4}"),
```
Comment (seelabs) [inline src/ripple/app/tests/TxQ_test.cpp]: Could use a raw string here to get rid of the backslashes: `R"("{"LastLedgerSequence":4})"`

**Comment by seelabs** [comment]
:+1:

File: src/ripple/app/tests/TxQ_test.cpp (hunk 2 above)
Comment (ximinez) [inline src/ripple/app/tests/TxQ_test.cpp]: I saw the same thing locally, and have already updated.

Comment (HowardHinnant) [inline src/ripple/app/tests/TxQ_test.cpp]: There is movement on the committee in this area.  I have seen proposals that would allow this to work without the `move`.  I _think_ they haven't been voted in yet; checking on that.  We are probably looking at vendors gaining field experience in this area to support such a proposal.

File: src/ripple/app/tests/TxQ_test.cpp (hunk 3 above)
Comment (ximinez) [inline src/ripple/app/tests/TxQ_test.cpp]: Done. The raw string takes one more byte in the source, ironically.

File: src/ripple/app/tests/TxQ_test.cpp (hunk 2 above)
Comment (ximinez) [inline src/ripple/app/tests/TxQ_test.cpp]: @HowardHinnant Good to know!

**Comment by ximinez** [comment]
Ping @nbougalis.

**Comment by ximinez** [comment]
Forgot to leave a comment when I did it, but pushed another commit moving the misplaced lock we discussed on Friday.

**Comment by seelabs** [comment]
:+1: On moving the lock. The only concern I had was if something called back into `TxQ` and tried to relock the mutex., but I didn't find anything.

**Comment by ximinez** [comment]
@seelabs Thanks for checking on that. I was careful to avoid recursion, but confirmation is the whole idea.

**Comment by nbougalis** [comment]
:+1:

**Comment by ximinez** [comment]
Rebased to 0.30.1-b7. No weirdness encountered.

**Comment by nbougalis** [comment]
Merged as 9c8204f945512fa7663495e2064fae5c9c62eeb9 and b87eff2115647cc4d8ed24b751f1077dfb99c326.
